Add NomialData class, reorganization

bqpd opened the pull request:
Closes #335.

bqpd commented:
Weird, it merges fine from here...

bqpd commented:
retest this please

bqpd commented:
retest this please

bqpd commented:
retest this please (doing this a lot)

galbramc commented:
I'm changing something with the script that is causing the failure. Give me a sec.

bqpd commented:
Sure thing!

galbramc commented:
test this please

galbramc commented:
We also need to get web hooks setup to that jenkins knows right away that something has changed. But we should work on that after you guys have finished this next release. @whoburg Probably should open an issue so we don't forget about it.

bqpd commented:
retest this please

bqpd commented:
`Build step 'Conditional step (single)' marked build as failure` hmmmm

galbramc commented:
Where did you see that error?

bqpd commented:
[Here](https://acdl.mit.edu/csi/job/gpkit/buildnode=reynolds,optimizer=mosek/306/console) near the bottom

galbramc commented:
It's the sed command I'm adding to only test mosek that is the problem. sed on Linux and Mac are apparently different.

bqpd commented:
[Also on macy's](https://acdl.mit.edu/csi/job/gpkit/buildnode=macys,optimizer=mosek/307/console)

galbramc commented:
test this please

bqpd commented:
Whoo, just macy's now!

galbramc commented:
It's the sed command. It works differently on Linux and OS X. I think I finally was able to get one command to work.

galbramc commented:
test this please

galbramc commented:
Finally passed. We are now testing cvxopt only, and mosek only. We don't test cvxopt+mosek. Mosek counts as two solvers though because of cli, so you are still testing multiple solvers.

bqpd commented:
Awesome!

bqpd commented:
@whoburg, this is ready for review.

whoburg commented:
roger. a lot moved! before I start, can you give a quick rundown of the major moves and reasoning, just to provide context?

bqpd commented:
Sure thing!Model.formProgram => it's own file
most Model properties => a model NomialData property
Model.separatesubs => substitution.py
Model.parse_result => solutionarray.py

created NomialData class
small_scripts.sort_and_simplify => nomial_data.py
locate_vars => nomial_data.py
Signomial.__hash__ => NomialData

nomials now inherit from NomialData

renamed varkeys dictionaries the more accurate varstrs

renamed substitution.getsubs to the more accurate substitution.getconstants
refactored substitutions to have nomialdata arguments instead of exps, cs, etc

bqpd commented:
retest this please

whoburg commented:
shoot, I forgot about this. Should I start code review now? Note merge conflicts should be fixed first if possible.

bqpd commented:
Sure! code review while I fix those?

bqpd commented:
Fixed.

whoburg commented:
starting now

bqpd commented:
retest this please

bqpd commented:
retest this please

whoburg commented on gpkit/form_program.py:
```diff
@@ -0,0 +1,59 @@
+
+from .substitution import substitution
+from .small_scripts import mag
+
```
Overall comment: I'm a bit critical of this function and whether it needs to exist -- detailed comments below.

whoburg replied:
isn't it more like "`apply_substitutions`"?

bqpd replied:
Yeah! And "collect mmaps"

bqpd replied:
Fixed!

whoburg commented on gpkit/form_program.py:
```diff
@@ -0,0 +1,59 @@
+
+
+def form_program(programType, signomials, subs, verbosity=2):
+    """Generates a program and solves it, sweeping as appropriate.
```
doesn't seem to solve or sweep...

bqpd replied:
Haha, good point.

bqpd replied:
Fixed!

whoburg commented on gpkit/form_program.py:
```diff
@@ -0,0 +1,59 @@
+    -------
+    program : GP or SP
+    mmaps : Map from initial monomials to substitued and simplified one.
+            See small_scripts.sort_and_simplify for more details.
```
since this already returns a tuple, what about having it return cost, constraints, mmaps, and leave out the last-step program creation?

That would eliminate `programType` and `verbosity`, and make this a function that does substitution on a list of Signomials. (move to `substitution.py`?)

bqpd replied:
Nice, I like that.

bqpd replied:
Done!

whoburg commented on gpkit/form_program.py:
```diff
@@ -0,0 +1,59 @@
+
+    verbosity : int (optional)
+        If greater than 0 prints runtime messages.
+        Is decremented by one and then passed to program inits.
```
seems to only have this behavior for gp, not sp.

bqpd replied:
Only gp has a verbosity arg in its init

bqpd replied:
Added verbosity to SP.

whoburg commented on gpkit/model.py:
```diff
@@ -130,80 +126,31 @@ def signomials(self):
-            variables.update({vk: Variable(**vk.descr) for vk in s.varlocs})
-        return variables
+    def unsubbed(self):
+        return NomialData(signomials=self.signomials, simplify=False)
```
wow, that worked out really nicely

whoburg replied:
the name unsubbed is slightly confusing, since this seems to be the NomialData of all signomials, whether or not substituted... ?

bqpd replied:
The `Model.substitutions` dict and the variable's values have not been applied to them. Open to new names, this convention's been in the code for ages.

whoburg replied:
I see. No, that does make sense.

whoburg commented on gpkit/model.py:
```diff
@@ -130,80 +126,31 @@ def signomials(self):
+        unsubbed = NomialData(signomials=signomials, simplify=False)
+        allsubs = unsubbed.values
+        allsubs.update(self.substitutions)
+        return signomials, unsubbed, allsubs
```
lol, I'll let this slide  :)

bqpd replied:
"For internal use only"

whoburg commented on gpkit/model.py:
```diff
@@ -338,8 +285,8 @@ def _solve(self, programType, solver, verbosity, skipfailures,
-        posynomials = self.signomials
-        sweep, linkedsweep, constants = self.separatesubs
+        signomials, unsubbed, allsubs = self.signomials_et_al
+        sweep, linkedsweep, constants = separatesubs(unsubbed, allsubs)
```
the name `separatesubs` is not very descriptive

bqpd replied:
It separates allsubs into constants, sweeps, and linkedsweeps...

whoburg replied:
I admit I don't have much better ideas... `bin_subs_by_type` ?

whoburg commented on gpkit/model.py:
```diff
@@ -367,16 +314,15 @@ def solve_pass(i):
                     if programType == "gp":
                         result = program.solve(*args, **kwargs)
                     elif programType == "sp":
                         result = program.localsolve(*args, **kwargs)
```
the conditional inside this `try` makes me further think that `form_program` should just return the parts needed to call the constructor (which would be done here).

bqpd replied:
It is done!

whoburg commented on gpkit/model.py:
```diff
@@ -388,24 +334,24 @@ def solve_pass(i):
-            sol = self.parse_result(result, constants, mmaps)
-            solution.append(sol)
-
+            solution.append_parse(result, constants, unsubbed)
```
"append_parse"? interesting

whoburg commented on gpkit/nomial_data.py:
```diff
@@ -0,0 +1,128 @@
+
+
+class NomialData(object):
+
```
needs a docstring

whoburg replied:
should the kwarg `signomials` be renamed `nomials` to make it clear it can be anything?

bqpd replied:
Ah, good call. It only needs cs and exps.

bqpd replied:
Done.

bqpd replied:
Done.

whoburg commented on gpkit/nomial_data.py:
```diff
@@ -0,0 +1,128 @@
+class NomialData(object):
+
+    def __init__(self, exps=None, cs=None, signomials=None, simplify=True):
+        if signomials:
```
should we `raise` here `if exps or cs`?

bqpd replied:
Done.

whoburg commented on gpkit/nomial_data.py:
```diff
@@ -0,0 +1,128 @@
+
+    def __init__(self, exps=None, cs=None, signomials=None, simplify=True):
+        if signomials:
+            exps = functools_reduce(add, (s.exps for s in signomials))
```
is this using functools_reduce to append tuples? That's a little confusing  :/

whoburg commented on gpkit/nomial_data.py:
```diff
@@ -0,0 +1,128 @@
+
+
+def sort_and_simplify(exps, cs, return_map=False):
+    """Reduces the number of monomials, and casts them to a sorted form.
```
After reading this first line I still don't know what this does

whoburg commented on gpkit/nomial_data.py:
```diff
@@ -0,0 +1,128 @@
+        The coefficients of each monomial
+    return_map : bool (optional)
+        Whether to return the map of which monomials combined to form a
+        simpler monomial, and their fractions of that monomial's final c.
```
hmph, my pylint likes this return conditional about as much as me (not a lot) -- though I see it's possibly efficient.

whoburg replied:
left-field idea -- to deal with the fact that `matches` computation is shared between exps/cs and maps

 - make `matches` a property of `NomialData` (with possibly an updated name). Implement by initializing matches attribute to `None` in `NomialData.__init__`, and recomputing matches only if thy are `None`.
 - Use matches to implement the computing of exps and cs from sort_and_simplify
 - separate mmaps function that also uses `matches`

bqpd replied:
I like where that'd going, but `matches` doesn't have enough information for that; it loses track of original indexes.

whoburg commented on gpkit/nomial_data.py:
```diff
@@ -0,0 +1,128 @@
+
+
+def locate_vars(exps):
+    "From exponents form a dictionary of which monomials each variable is in."
```
I haven't carefully examined the logic in this method  -- should I look through it carefully?

whoburg replied:
more inline comments would be really helpful with all the conditionals.

bqpd replied:
It's old and gross, yeah.

whoburg replied:
done, #352

whoburg commented on gpkit/nomials.py:
```diff
@@ -25,9 +24,6 @@ class Signomial(object):
             Coefficient values for each monomial term
-        varlocsandkeys: dict
-            mapping from variable name to list of indices of monomial terms
-            that variable appears in
```
finally that horribly-named thing dies.

whoburg commented on gpkit/nomials.py:
```diff
@@ -100,11 +94,12 @@ def __init__(self, exps=None, cs=1, varlocsandkeys=None,
+            any_negative = any(c.magnitude <= 0 for c in self.cs)
         else:
-            any_negative = any((c <= 0 for c in cs))
+            any_negative = any(c <= 0 for c in self.cs)
```
should this go in `small_scripts.mag`?

bqpd replied:
Oh, I like that idea.

bqpd replied:
Done.

whoburg commented on gpkit/nomials.py:
```diff
@@ -112,29 +107,20 @@ def __init__(self, exps=None, cs=1, varlocsandkeys=None,
-            units = cs[0]/cs[0].magnitude
+        units = None
+        if isinstance(self.cs[0], Quantity):
+            units = self.cs[0]/self.cs[0].magnitude
```
ugh, pint doesn't have a nicer way of getting units?

bqpd replied:
There's `Q.units`, but that isn't multipliable or to-able. :-/

whoburg commented on gpkit/nomials.py:
```diff
@@ -226,13 +207,9 @@ def prod(self):
     def __ne__(self, other):
         if isinstance(other, Signomial):
-            return not (self.exps == other.exps and self.cs == other.cs)
+            return hash(self) != hash(other)
```
nice nice nice

whoburg commented on gpkit/nomials.py:
```diff
@@ -287,10 +264,6 @@ def __str__(self, mult_symbol='*'):
-    def descr(self, descr):
-        self.descr = descr
-        return self
-
```
thank goodness, this always confused me

whoburg commented on gpkit/signomial_program.py:
```diff
@@ -108,8 +109,11 @@ def localsolve(self, solver=None, verbosity=1, x0=None, reltol=1e-4,
+            self.negydata = lambda: None
+            self.negydata.varlocs = self.negvarkeys
+            self.negydata.varstrs = {str(vk): vk for vk in self.negvarkeys}
+            x0 = getconstants(self.negydata, x0)
```
temporary, deserves # HACK comment?

bqpd commented on gpkit/signomial_program.py:
```diff
@@ -108,8 +112,11 @@ def localsolve(self, solver=None, verbosity=1, x0=None, reltol=1e-4,
         else:
-            x0 = getsubs({str(vk): vk for vk in self.negvarkeys},
-                         self.negvarkeys, x0)
+            # dummy nomial data to turn x0's keys into VarKeys
```
unclear if this is a temporary hack or not. :p

whoburg commented on gpkit/solution_array.py:
```diff
@@ -29,16 +34,27 @@ class SolutionArray(DictOfLists):
+        *args, **kwargs : passed to parse_result()
+        """
+        self.append(parse_result(*args, **kwargs))
+
```
um, is this better than just calling the commands in series? It's certainly more obfuscated.

bqpd replied:
I thought it'd be nice to users who wanted to make their own solution array
(i.e. if they were writing their own sweep?) but maybe not.

bqpd replied:
Removed.

whoburg commented on gpkit/substitution.py:
```diff
@@ -70,7 +70,40 @@ def vectorsub(subs, var, sub, varset):
 
 
-def substitution(varlocs, varkeys, exps, cs, substitutions, val=None):
+def separatesubs(nomial, substitutions):
```
hmm, this seems to mostly do things `if is_sweepvar(sub)`. should it be renamed?

bqpd replied:
That's because all the other goodness is in get_constants. They should be
combined better, though.

whoburg commented:
code review complete; haven't looked at responses yet. additional suggestions in #351

bqpd commented:
retest this please

bqpd commented:
I think the Jenkins results occasionally get returned for the next test by accident (when there's multiple commit, maybe?)

bqpd commented:
Discuss: `super` or `NomialData`? (re: #337 )

galbramc commented:
@bqpd I checked that when I pushed multiple commits in just one push, jenkins only tested the most recent commit, not all the commits I pushed. Is it still testing all the commits for you?

bqpd commented:
I dunno. It just failed to fail on a bad commit (5c9b765) and confused me.

whoburg commented:
Re: super, I'm all for GP and Signomials inheriting from NomialData, and indeed it's nice that we get direct access to `cs` and `exps` that way.

I've never heard of not calling `super` for inheritance. But I think they end up doing almost exactly the same thing -- I don't know of a technical reason to argue for one over the other. I just know that people use super.

galbramc commented:
Looks like there is a bug in jenkins. The commit for the individual builds are not matching the overall build. I had to fix a problem just like this when I first started using jenkins, but it was in a different plugin. I have to dig into it to see what I can find...

galbramc commented:
test this please

galbramc commented:
As best as I can tell, the testing gets out of sync when there is a merge conflict... The last couple of tests were actually consistently testing one commit older than the most recent. This last test request put things back in sync again. I think I need to get in touch with the plugin developers to resolve this one.

galbramc commented:
You are back in sync now btw.

bqpd commented:
Great, thanks!

whoburg commented:
what's our timeline on merging this? If it's soon I'll wait to start my branch for work on #23 from the post-merge master

bqpd commented:
The sooner the better, definitely. I'd rather merge this now with `super` than wait.

whoburg commented:
It's also related to #307 -- I think I'm arguing for super in part because we're definitely doing this wrong in other parts of the code, and getting all inheritance to use super would resolve that

bqpd commented:
I'm dubious about that, but I'll make the change :p

whoburg commented:
or if you prefer, you can leave `NomialData` in place but at least leave an inline comment saying that it's implementing super.

bqpd commented:
I added an explanatory comment, I think it'll work out.

bqpd commented:
@whoburg, Ready for merging!

galbramc commented:
test this please
It's behind again!?! It actually tested commit c6586aa

whoburg commented:
@galbramc let me know when you think this is tested -- otherwise ready to merge.

bqpd commented:
Conflicts after last PR! Fixing

bqpd commented:
retest this please

bqpd commented:
(looks like it was still one behind; it failed to merge)
